promql/functions.go

```diff
@@ -1426,23 +1426,75 @@ func funcResets(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelpe
 	return append(enh.Out, Sample{F: float64(resets)}), nil
 }
 
+// mergeSeries merges the floats and the histograms in a series
+// into a slice of Sample sorted by their timestamp.
+// it assumes that the floats and histograms are sorted by their timestamp.
+func mergeSeries(series Series) []*Sample {
+	floats := series.Floats
+	histograms := series.Histograms
+	samples := make([]*Sample, 0, len(floats)+len(histograms))
+
+	// i for iterating through floats and j for iterating through histograms
+	i, j := 0, 0
+	for i < len(floats) && j < len(histograms) {
+		if floats[i].T < histograms[j].T {
+			samples = append(samples, &Sample{
+				F: floats[i].F,
+			})
+			i++
+		} else {
+			samples = append(samples, &Sample{
+				H: histograms[j].H,
+			})
+			j++
+		}
+	}
+	// Append the remaining samples
+	for i < len(floats) {
+		samples = append(samples, &Sample{
+			F: floats[i].F,
+		})
+		i++
+	}
+	for j < len(histograms) {
+		samples = append(samples, &Sample{
+			H: histograms[j].H,
+		})
+		j++
+	}
+
+	return samples
+}
+
 // === changes(Matrix parser.ValueTypeMatrix) (Vector, Annotations) ===
 func funcChanges(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper) (Vector, annotations.Annotations) {
-	floats := vals[0].(Matrix)[0].Floats
+	samples := mergeSeries(vals[0].(Matrix)[0])
```

Comment:
As the series can actually become rather long (users can write things like `changes(foobar[1y])`), I would say we should try really hard to avoid the creation of a new `samples` slice.

At the very least, we could test if the series is really mixed and do the old thing when it is not.

However, I think it should be fairly easy to avoid `mergeSeries` altogether, doing something like

```suggestion
	floats := vals[0].(Matrix)[0].Floats
	histograms := vals[0].(Matrix)[0].Histograms
	for iFloats, iHistograms := 0; iFloats < len(floats) && iHistograms < len(histograms); {
	     // Pick the earlier sample between floats[iFloats] and histograms[iHistograms] (if any).
	     // Store it as `Sample`.
	     // Increment the respective iFloats or  iHistograms.
	     // Compare the stored `Sample` to the previous (if any) and increment `changes` accordingly.
	     // Make the stored `Sample` the previous sample.
	}
```

(Might be a bit fiddlier than sketched out above in detail, but I hope you get what I mean.)

Reply:
Thank you, @beorn7, for your input! Initially, I also considered comparing them on the fly, but I shifted to this approach since the same functionality would be utilized in `resets` as well. However, after running benchmarks, the results were not promising. Given the high memory consumption of this method, I believe it's more efficient to perform the comparisons on the go. I'll shift to the "on the fly" approach.